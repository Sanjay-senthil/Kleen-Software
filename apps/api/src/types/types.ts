import {
  AccessControlPermission,
  Axis,
  EntityList,
  FilterQuery,
  GetListingDataResults,
  GetWidgetDataResult,
  ListItemBase as ListItem,
  ListingFormatItem,
  ListingItem,
  MultiTransFormationResults,
  Transformation,
  WorkflowFilter,
} from '@kleeen/types';

// Duplicated type, must unify in types library
export interface Attribute {
  aggregation?: Transformation;
  name: string;
  // TODO @carreta remove this when XORs can be aggregated [KSE3-1735]
  isXor?: boolean;
  hasMany?: boolean;
  formatType?: string;
}

export interface WorkflowFiltersResults {
  filters: WorkflowFilter[];
}

interface ActionItem {
  link?: string;
  title?: string;
  type?: string;
}

// TODO: @cafe move this and other shared types to @kleeen/types
export interface DispatchCustomActionResults {
  data: {
    actions?: ActionItem[];
    context?: any;
    customMessage?: string;
    customTitle?: string;
    functionName: string;
    success: boolean;
  };
}

export interface MultiTransFormationArgs {
  attributes?: string[];
  entity: string;
  filters: FilterQuery;
  transformations: Transformation[];
}

export interface FormatCheckResults {
  isValid: boolean;
  errors?: {
    message: string;
  }[];
}

export interface AccessControlCheckResults {
  accessLevel: AccessControlPermission;
}

export interface AuthContext {
  dataSources: Record<string, any>;
  headers: Record<string, any>;
  user: {
    email: string;
    'cognito:id': string;
  };
  token: string;
}

export {
  Axis,
  EntityList,
  GetListingDataResults,
  GetWidgetDataResult,
  ListingFormatItem,
  ListingItem,
  ListItem,
  MultiTransFormationResults,
  Transformation,
};
