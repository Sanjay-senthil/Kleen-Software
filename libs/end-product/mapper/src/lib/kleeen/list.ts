import { ColumnMap, DataPointValue, GetListingDataResults, MapFunctionType } from '@kleeen/types';
import { toChipsValue, toDisplayValue, toIdValue } from '@kleeen/end-product/utils';

import { getThingByName } from '@kleeen/things';
import { isNilOrEmpty } from '@kleeen/common/utils';

interface Value {
  [key: string]: DataPointValue | DataPointValue[];
}

const handlerByType = {
  [MapFunctionType.ToChipsValue]: toChipsValue,
  [MapFunctionType.ToDisplayValue]: toDisplayValue,
  [MapFunctionType.ToIdValue]: toIdValue,
};

export function mapToList(rows: any[], columns: ColumnMap[]): GetListingDataResults {
  const data = mapListColumns(rows, columns);
  const reducedFormats = columns.reduce((acc, column) => {
    const thingFormat = getThingByName(column.key)?.format;

    if (isNilOrEmpty(thingFormat)) {
      return acc;
    }

    return {
      ...acc,
      [column.key]: thingFormat,
    };
  }, {});

  return {
    data,
    format: reducedFormats,
  };
}

export function mapListColumns(rows: any[], columns: ColumnMap[], reduceColumns = true) {
  const columnsMap = mapColumns(columns);

  if (Array.isArray(rows)) {
    const values = rows.map((row) => mapListColumns(row, columns, reduceColumns));

    return values;
  }

  const mappedColumns = Object.values(columnsMap)
    .map((columnMapper: (rows: unknown) => Value) => {
      if (typeof columnMapper == 'function') {
        return columnMapper(rows);
      }

      return null;
    })
    .filter(Boolean);

  if (!reduceColumns) return mappedColumns;

  return mappedColumns.reduce(
    (previous, current) => ({
      ...previous,
      ...current,
    }),
    {},
  );
}

// #region Private members
function mapColumns(columns: ColumnMap[]) {
  return columns.reduce(
    (acc, { handler = MapFunctionType.ToDisplayValue, idPath, key, path, showDisplayMedia = true }) => {
      return {
        ...acc,
        [path]: (row) => {
          const columnHandler = handlerByType[handler];

          return columnHandler({
            idPath,
            key,
            path,
            row,
            showDisplayMedia,
          });
        },
      };
    },
    {},
  );
}
// #endregion
