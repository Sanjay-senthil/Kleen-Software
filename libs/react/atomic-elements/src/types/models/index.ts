import { CrossLinking, FormatProps, ResultsProps } from '@kleeen/types';

export interface ContextProps {
  data: DataProps;
  error: string;
  isLoading: boolean;
}

export interface DataProps extends ResultsProps {
  crossLinking: CrossLinking[] | CrossLinking;
  format: FormatProps;
}

export interface XAxisOptions extends Highcharts.XAxisOptions {
  categories?: [string];
  key?: string;
  valueLabels?: { label: string; value: number }[];
}

export interface YAxisOptions extends Highcharts.YAxisOptions {
  categories?: [string];
  key?: string;
  valueLabels?: { label: string; value: number }[];
}
