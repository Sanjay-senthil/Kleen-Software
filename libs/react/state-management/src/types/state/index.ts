export * from './endUser';
export * from './globalAppErrors';
export * from './investigation';
