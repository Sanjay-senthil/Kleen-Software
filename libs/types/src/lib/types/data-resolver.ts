import {
  Bucket,
  CrossLinking,
  CrossLinkingMatrix,
  DisplayValue,
  Transformation,
  UpdateDataStrategy,
} from '@kleeen/types';

/**
 * Enumerations
 */

export enum MapFunctionType {
  ToDisplayValue = 'toDisplayValue',
  ToChipsValue = 'toChipsValue',
  ToIdValue = 'toIdValue',
}

/**
 * Interfaces
 */

export interface Axis {
  categories?: string[] | ListItemBase[];
  key?: string;
  type?: string;
}

export interface ColumnMap {
  handler?: MapFunctionType;
  idPath?: string;
  key: string;
  path: string;
  showDisplayMedia?: boolean;
}

export interface Entity {
  name: string;
  properties: { [property: string]: { statisticalType: string } };
}

export interface EntityList {
  [key: string]: ListingItem | ListingItem[];
}

export interface DataAggregationArgsDataPoint {
  bucket?: Bucket;
  name: string;
  transformation?: Transformation;
  transformations?: Transformation[];
}

export interface GetListingDataResults {
  data: EntityList[];
  format: { [key: string]: ListingFormatItem };
  latestRequestTimestamp?: number;
  pagination?: { totalCount: number };
  strategy?: UpdateDataStrategy;
}

export interface GetWidgetDataResult {
  crossLinking?: CrossLinkingMatrix;
  format: {
    xAxis?: Axis;
    yAxis?: Axis;
  };
  results: number[] | number[][];
}

export interface ListingFormatItem {
  aggregations?: null;
  examples?: null;
  max?: null;
  min?: null;
  prefix?: null;
  severityBad?: null;
  severityGood?: null;
  severityLevels?: null;
  suffix?: null;
}

export interface ListingItem {
  displayValue: DisplayValue;
  id?: string;
}

export interface MultiTransFormationResults {
  crossLinking?: CrossLinking[];
  format: {
    xAxis?: Axis;
    yAxis?: Axis;
  };
  results: number | number[] | number[][];
  transformation: Transformation;
}

export interface TransformationArgs extends Pick<ColumnMap, 'idPath' | 'key' | 'path' | 'showDisplayMedia'> {
  row: any;
}

/**
 * Types
 */

export type ListItemBase = number | string | boolean | undefined;
