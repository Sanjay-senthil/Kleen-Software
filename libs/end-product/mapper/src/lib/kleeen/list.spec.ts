import { MapFunctionType, Maybe, Thing } from '@kleeen/types';

import { mapToList } from './list';
import { thingsMock } from './mocks';

jest.mock('@kleeen/things', () => {
  return {
    getThingByName(name: string): Maybe<Thing> {
      return Object.values(thingsMock).find((thing) => thing.name === name) as Thing;
    },
  };
});

describe('kleeen', () => {
  describe('list', () => {
    it('should return "GetListingDataResults" response if "mapToList" is called from a Full Table widget', () => {
      const columns = [
        {
          idPath: 'Vehicle.vehicleId',
          key: 'id',
          handler: MapFunctionType.ToIdValue,
          path: 'id',
        },
        {
          idPath: 'Vehicle.vehicleId',
          key: 'vehicle',
          path: 'Vehicle.vehicleId',
          showDisplayMedia: false,
        },
        {
          key: 'vehicleBrand',
          path: 'Vehicle.brand',
          showDisplayMedia: false,
        },
        {
          key: 'vehicleModel',
          path: 'Vehicle.model',
          showDisplayMedia: false,
        },
      ];
      const rows = [
        {
          'Vehicle.model': 'S2',
          'Vehicle.brand': 'ByD',
          'Vehicle.vehicleId': '1e9aa1aaxkd630001',
        },
        {
          'Vehicle.model': 'Lifestyle Vehicle',
          'Vehicle.brand': 'Canoo',
          'Vehicle.vehicleId': '1e9aa1aaxkd630002',
        },
        {
          'Vehicle.model': 'S',
          'Vehicle.brand': 'Tesla',
          'Vehicle.vehicleId': '1e9aa1aaxkd630003',
        },
        {
          'Vehicle.model': 'Prius',
          'Vehicle.brand': 'Toyota',
          'Vehicle.vehicleId': '1e9aa1aaxkd630004',
        },
      ];

      expect(mapToList(rows, columns)).toMatchSnapshot();
    });

    it('should return "GetListingDataResults" response if "mapToList" is called from a Full Table widget with Chips', () => {
      const columns = [
        {
          key: 'vehicleEcu',
          path: 'DTC.ecu',
          showDisplayMedia: false,
        },
        { key: 'vehicleSoftwareVersion', path: 'DTC.version', showDisplayMedia: false },
        {
          handler: MapFunctionType.ToChipsValue,
          idPath: 'DTC.dtc',
          key: 'vehicleDtc',
          path: 'DTC.dtcText',
        },
      ];

      const rows = [
        {
          'DTC.ecu': 'BPCM',
          'DTC.version': '00.40.01',
          'DTC.dtcText': [
            {
              'DTC.dtc': 'B1006-19',
              'DTC.dtcText':
                'Fan Interior Ambient Temperature Sensor Front Over Current - Circuit Current Above Threshold',
            },
            {
              'DTC.dtc': 'B1033-16',
              'DTC.dtcText': 'Side view Mirror Driver Under Voltage L/R - Circuit Voltage Below Threshold',
            },
            {
              'DTC.dtc': 'B1028-19',
              'DTC.dtcText': 'Hazard Light Switch Over Current - Circuit Current Above Threshold',
            },
          ],
        },
      ];

      expect(mapToList(rows, columns)).toMatchSnapshot();
    });

    it('should return "GetListingDataResults" response if "mapToList" is called from a Full Table widget with Severity Scores', () => {
      const rows = [
        {
          'Vehicle.stateOfCharge': '11.5',
          'Vehicle.timestamp': '1643917172',
          'Vehicle.vehicleId': '1e9aa1aaxkd630008',
        },
        {
          'Vehicle.stateOfCharge': '44.5',
          'Vehicle.timestamp': '1643917174',
          'Vehicle.vehicleId': '1e9aa1aaxkd630008',
        },
        {
          'Vehicle.stateOfCharge': '67.5',
          'Vehicle.timestamp': '1643917175',
          'Vehicle.vehicleId': '1e9aa1aaxkd630008',
        },
        {
          'Vehicle.stateOfCharge': '93.5',
          'Vehicle.timestamp': '1643917176',
          'Vehicle.vehicleId': '1e9aa1aaxkd630008',
        },
      ];
      const columns = [
        {
          idPath: 'Vehicle.vehicleId',
          key: 'vehicle',
          path: 'Vehicle.vehicleId',
          showDisplayMedia: false,
        },
        { key: 'vehicleStateOfCharge', path: 'Vehicle.stateOfCharge' },
        { key: 'vehicleTimeOfLastCharge', path: 'Vehicle.timestamp', showDisplayMedia: false },
      ];

      expect(mapToList(rows, columns)).toMatchSnapshot();
    });
  });
});
