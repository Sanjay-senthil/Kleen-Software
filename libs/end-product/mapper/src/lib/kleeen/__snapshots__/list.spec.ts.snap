// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kleeen list should return "GetListingDataResults" response if "mapToList" is called from a Full Table widget 1`] = `
Object {
  "data": Array [
    Object {
      "id": "1e9aa1aaxkd630001",
      "vehicle": Object {
        "displayValue": "1e9aa1aaxkd630001",
        "id": "1e9aa1aaxkd630001",
      },
      "vehicleBrand": Object {
        "displayValue": "ByD",
        "id": undefined,
      },
      "vehicleModel": Object {
        "displayValue": "S2",
        "id": undefined,
      },
    },
    Object {
      "id": "1e9aa1aaxkd630002",
      "vehicle": Object {
        "displayValue": "1e9aa1aaxkd630002",
        "id": "1e9aa1aaxkd630002",
      },
      "vehicleBrand": Object {
        "displayValue": "Canoo",
        "id": undefined,
      },
      "vehicleModel": Object {
        "displayValue": "Lifestyle Vehicle",
        "id": undefined,
      },
    },
    Object {
      "id": "1e9aa1aaxkd630003",
      "vehicle": Object {
        "displayValue": "1e9aa1aaxkd630003",
        "id": "1e9aa1aaxkd630003",
      },
      "vehicleBrand": Object {
        "displayValue": "Tesla",
        "id": undefined,
      },
      "vehicleModel": Object {
        "displayValue": "S",
        "id": undefined,
      },
    },
    Object {
      "id": "1e9aa1aaxkd630004",
      "vehicle": Object {
        "displayValue": "1e9aa1aaxkd630004",
        "id": "1e9aa1aaxkd630004",
      },
      "vehicleBrand": Object {
        "displayValue": "Toyota",
        "id": undefined,
      },
      "vehicleModel": Object {
        "displayValue": "Prius",
        "id": undefined,
      },
    },
  ],
  "format": Object {
    "vehicle": Object {
      "examples": Array [
        "Fleet Car 1",
        "Fleet Car 2",
        "Fleet Car 3",
        "Fleet Car 4",
        "Loaner Car",
        "Loaner Van",
        "Loaner Truck",
        "Service Van 1",
        "Service Van 2",
        "Delivery Van",
        "Delivery Truck",
        "Service Truck 1",
        "Service Truck 2",
        "Executive Car 1",
        "Executive Car 2",
        "Executive Car 3",
        "Sales Car 1",
        "Sales Car 2",
        "Sales Car 3",
      ],
    },
    "vehicleBrand": Object {
      "examples": Array [
        "Datsun",
        "Delorean",
        "Pontiac",
        "Plymouth",
        "Desoto",
        "Mercury",
        "Tucker",
        "Oldsmobile",
        "Studebaker",
        "Kaiser",
        "Saturn",
        "AMC",
        "Edsel",
        "Auburn",
        "Checker",
        "Dusenburg",
        "Fisker",
        "Hudson",
        "MGB",
        "Triumph",
      ],
    },
    "vehicleModel": Object {
      "examples": Array [
        "M1",
        "M2",
        "M3",
        "M5",
        "M4",
        "S1",
        "S2",
        "S3",
        "S4",
        "S5",
        "X1",
        "X2",
        "X3",
        "X4",
        "X5",
        "C1",
        "C2",
        "C3",
        "C4",
        "C5",
      ],
    },
  },
}
`;

exports[`kleeen list should return "GetListingDataResults" response if "mapToList" is called from a Full Table widget with Chips 1`] = `
Object {
  "data": Array [
    Object {
      "vehicleDtc": Array [
        Object {
          "displayValue": "Fan Interior Ambient Temperature Sensor Front Over Current - Circuit Current Above Threshold",
          "id": "B1006-19",
        },
        Object {
          "displayValue": "Side view Mirror Driver Under Voltage L/R - Circuit Voltage Below Threshold",
          "id": "B1033-16",
        },
        Object {
          "displayValue": "Hazard Light Switch Over Current - Circuit Current Above Threshold",
          "id": "B1028-19",
        },
      ],
      "vehicleEcu": Object {
        "displayValue": "BPCM",
        "id": undefined,
      },
      "vehicleSoftwareVersion": Object {
        "displayValue": "00.40.01",
        "id": undefined,
      },
    },
  ],
  "format": Object {
    "vehicleDtc": Object {
      "examples": Array [
        "Air Cleaning Module Overcurrent - Circuit Current Above Threshold",
        "Backlite Defrost Relay Fault - Signal Shape / Waveform Failure",
        "Cabin Heater (PTCHF) 12V Logic Over Voltage - Circuit Voltage Above Threshold",
        "Door Handle Rear Left 12V Supply Under Voltage - Circuit Voltage Below Threshold",
        "Ext Door Handle Power RR Undercurrent - Circuit Current Below Threshold",
        "External 12VA Out of Range - Circuit Voltage Out of Range",
      ],
    },
    "vehicleEcu": Object {
      "examples": Array [
        "RIDM",
        "BPCU",
        "FIDM",
        "RBCM",
        "FBCM",
        "DCAM",
        "DCBM",
        "OBCMD",
        "OBCMP",
      ],
    },
    "vehicleSoftwareVersion": Object {
      "examples": Array [
        "5.4.0",
        "5.4.1",
        "5.4.2",
        "5.5.0",
        "6.0.0",
      ],
    },
  },
}
`;

exports[`kleeen list should return "GetListingDataResults" response if "mapToList" is called from a Full Table widget with Severity Scores 1`] = `
Object {
  "data": Array [
    Object {
      "vehicle": Object {
        "displayValue": "1e9aa1aaxkd630008",
        "id": "1e9aa1aaxkd630008",
      },
      "vehicleStateOfCharge": Object {
        "displayMedia": null,
        "displayValue": "11.5",
        "id": undefined,
      },
      "vehicleTimeOfLastCharge": Object {
        "displayValue": "1643917172",
        "id": undefined,
      },
    },
    Object {
      "vehicle": Object {
        "displayValue": "1e9aa1aaxkd630008",
        "id": "1e9aa1aaxkd630008",
      },
      "vehicleStateOfCharge": Object {
        "displayMedia": null,
        "displayValue": "44.5",
        "id": undefined,
      },
      "vehicleTimeOfLastCharge": Object {
        "displayValue": "1643917174",
        "id": undefined,
      },
    },
    Object {
      "vehicle": Object {
        "displayValue": "1e9aa1aaxkd630008",
        "id": "1e9aa1aaxkd630008",
      },
      "vehicleStateOfCharge": Object {
        "displayMedia": null,
        "displayValue": "67.5",
        "id": undefined,
      },
      "vehicleTimeOfLastCharge": Object {
        "displayValue": "1643917175",
        "id": undefined,
      },
    },
    Object {
      "vehicle": Object {
        "displayValue": "1e9aa1aaxkd630008",
        "id": "1e9aa1aaxkd630008",
      },
      "vehicleStateOfCharge": Object {
        "displayMedia": null,
        "displayValue": "93.5",
        "id": undefined,
      },
      "vehicleTimeOfLastCharge": Object {
        "displayValue": "1643917176",
        "id": undefined,
      },
    },
  ],
  "format": Object {
    "vehicle": Object {
      "examples": Array [
        "Fleet Car 1",
        "Fleet Car 2",
        "Fleet Car 3",
        "Fleet Car 4",
        "Loaner Car",
        "Loaner Van",
        "Loaner Truck",
        "Service Van 1",
        "Service Van 2",
        "Delivery Van",
        "Delivery Truck",
        "Service Truck 1",
        "Service Truck 2",
        "Executive Car 1",
        "Executive Car 2",
        "Executive Car 3",
        "Sales Car 1",
        "Sales Car 2",
        "Sales Car 3",
      ],
    },
    "vehicleStateOfCharge": Object {
      "max": 100,
      "min": 0,
      "severityBad": 0,
      "severityGood": 100,
      "severityLevels": 5,
    },
    "vehicleTimeOfLastCharge": Object {
      "examples": Array [
        "02/23/2022, 4:28:57 PM",
        "05/23/2022, 4:28:57 PM",
      ],
    },
  },
}
`;
